src/qonnx/transformation/channelsLast.py

```diff
@@ -0,0 +1,460 @@
+
+                    # Attach to original node
+                    n.input[i] = inp_trans_out
+
+                # Insert transformation nodes for output nodes
+                output_tensors = n.output
+                for i, outp in enumerate(output_tensors):
+                    chanFirst_shape = model.get_tensor_shape(outp)
+                    ndim = len(chanFirst_shape)
+                    assert ndim == 3 or ndim == 4, "Channels last conversion is only available for 3D and 4D tensors."
+                    chanLast_shape = [chanFirst_shape[idx] for idx in _to_chan_last_args[ndim]]
+                    # Intermediat tensor
+                    outp_trans_in = helper.make_tensor_value_info(
+                        model.make_new_valueinfo_name(),
+                        TensorProto.FLOAT,
+                        chanLast_shape,
+                    )
+                    graph.value_info.append(outp_trans_in)
+                    outp_trans_in = outp_trans_in.name
+
+                    # ChannelsFirst -> ChannelsLast transpose
+                    outp_trans_node = helper.make_node("Transpose", [outp_trans_in], [outp], perm=_to_chan_first_args[ndim])
+                    graph.node.insert(running_node_index, outp_trans_node)
+                    running_node_index += 1
+
+                    # Attach to original node
+                    n.output[i] = outp_trans_in
+
+                # Modify domain
+                n.domain = "qonnx.custom_op.ChannelsLast"
+                # Set modified flag
+                graph_modified = True
+
+        return model, graph_modified
+
+
+class RemoveConsecutiveChanFirstAndChanLastTrafos(Transformation):
+    """
+    Remove two consecutive transformations, which would do:
+    (ChannelsLast -> ChannelsFirst) -> (ChannelsFirst -> ChannelsLast)
+    Or more concrete, the first converts to channels first and the second to channels last.
+    """
+
+    def apply(self, model):
+        graph = model.graph
+        node_ind = 0
+        graph_modified = False
+
+        # Find fitting transpose node pairs
+        for n in graph.node:
+            node_ind += 1
+            if n.op_type == "Transpose":
+                # Check the input shape and make sure we support it
+                input_shape = model.get_tensor_shape(n.input[0])
+                ndim = len(input_shape)
+                if ndim not in _to_chan_first_args.keys():
+                    continue
+                # Check that this is a "to chan first" trafo
+                perm_1 = get_by_name(n.attribute, "perm")
+                if list(_to_chan_first_args[ndim]) == perm_1.ints:
+
+                    successor_nodes = model.find_direct_successors(n)
+                    assert (
+                        len(successor_nodes) == 1
+                    ), "Transpose nodes should only have one output, I don't think more than one would even be possible."
+                    successor_node = successor_nodes[0]
```

Comment:
If the `Transpose` node was one inserted by `InsertChannelsLastDomainsAndTrafos`, and the node it was inserted after had multiple successors (e.g. a model with branches), could the `Transpose` node have multiple successors?

Reply:
That was my thinking, but after taking a look at a ResNet model ( https://github.com/onnx/models/blob/master/vision/classification/resnet/model/resnet50-v1-12.onnx ) it seems that this is shouldn't be of concern. Because a branch connection is implemented by connecting the same tensor to multiple nodes. This would then mean that the Transpose node still only has one output. The documentation also defines no optional inputs or outputs: https://github.com/onnx/onnx/blob/master/docs/Operators.md#Transpose
So I think it should be save to remove this statement.
An issue we could run into in the future is that people might use the optional outputs of the MaxPool and BatchNorm nodes, see: https://github.com/onnx/onnx/blob/master/docs/Operators.md#MaxPool and https://github.com/onnx/onnx/blob/master/docs/Operators.md#BatchNormalization
But at least for BatchNorm they are only used for training, which we don't support anyways. And I've also never seen a model make use of the optional outputs.

Reply:
Hi,
I'll give the ResNet ONNX file from the ONNX model zoo a try and check what happens.

Reply:
Okay, adding support for this appears to be a bit more involved. So I've added a check a the beginning of the transformation to check that a model is linear before it gets processed.